Dask blog post (#23)
JDASoftwareGroup/jdatech.github.io: _posts/2020-03-24-dask-developer-workshop.markdown

---
layout: single

> title:  "Tales from D.C.: Dask Developer Workshop 2020"

Author: the latest title is `Karlsruhe to D.C. ― a Dask story`

> title:  "Tales from D.C.: Dask Developer Workshop 2020"

Author: the tilte you proprose is great, but we want to include the element of _story-telling_ as we are talking about our experience with dask around the workshop

> title:  "Tales from D.C.: Dask Developer Workshop 2020"

Reviewer: Ok for me, but please consider, you have this one title to convince people to read or not read the blog post...if it is not 110% catchy you might miss a lot readers

date:   2020-03-24 20:00:00 +0100
tags: technology python data-engineering dask distributed
header:
  overlay_image: assets/images/tech_gear_banner.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false
author: Lucas Rademaker
author_profile: true
---
# Tales from D.C.: Dask Developer Workshop 2020

> 

Author: we have an introduction at `## The Blue Yonder way of using Dask`

The atmosphere of the workshop was great, people were eager to share their experiences with Dask and the ecosystem surrounding it. This provided for some interesting discussions.

### Final remarks
All in all, my colleagues and I found this workshop very valuable. We found this value in being able to share our issues and discuss potential ways of improvement. Additionally, listening to other Dask users and developers gave us a wider perspective on the current status of Dask, and the ecosystem around it, and what we could expect in the future; this allows us to focus on what's most efficient for us now.

Personally, I cannot say it enough times: I loved the working sessions. I learned a lot about Dask and distributed internals and made my first contribution to distributed when I sat down with [John Lee](https://github.com/leej3/) and we made it possible to benchmark work stealing for tasks with restrictions.

Apart from increasing my abilities to contribute upstream, this newly acquired knowledge about Dask has enabled me to aid my colleagues performing data science more effectively when they run into issues with these tools.

> Although I secretly hoped that with all these smart people in one room and all the time allocated to working sessions, _all_ major Dask issues would be resolved, I believe a good number of PR's were made during these days, and a lot of valuable knowledge shared which will provide its fruits in the mid-term.

Author: already removed AFAIK

---
layout: single
title:  "Tales from D.C.: Dask Developer Workshop 2020"
date:   2020-03-24 20:00:00 +0100
tags: technology python data-engineering dask distributed
header:
  overlay_image: assets/images/tech_gear_banner.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false
author: Lucas Rademaker
author_profile: true
---
# Tales from D.C.: Dask Developer Workshop 2020

## Setting
Back in January this year, [Dask](https://dask.org/) community members received an invite for a workshop in Washington DC with the goal of "connecting advanced users with developers".

> ## The workshop

Reviewer 2: Maybe you can merge the `Setting` and the `The workshop` section to one? Both feel a bit like an introduction.

The schedule of the workshop alone was quite satisfying to me: a number of talks, involving users and developers of Dask and related projects (e.g. NumPy, xarray, Arrow), combined with working sessions, wherein one had the opportunity to discuss or code with other attendees.

What follows is an outline of my experience at the workshop.
### Talks
##### ETL Pipelines
 - Blue Yonder and [Kartothek](https://github.com/JDASoftwareGroup/kartothek).
 

>     My colleague Florian Jetter was the first to present. He outlined our use case of executing Dask pipelines coupled with tight SLA's (Service Level Agreements), which obliges us to make sure that our Dask pipelines finish successfully each run, as usually we do not have time to re-schedule. He also talked about some of the issues we had faced in the past and others which we are facing in the present (mostly related to stability).

Reviewer 2: I would be interested in more details here, like what exactly is our use case for these dask pipelines? Which issues are we facing?
Obviously, we don't need a full summary of the talk. But just mentioning that we "faced issues" without actually naming what thise issues were feels a bit unsatisfactory.

- Prefect

    TODO

##### Deployment
  - dask-gateway

    TODO

  - dask-jobqueue

    TODO

##### Other kids on the block
 - xarray
 
    TODO

 - RAPIDS
 
    TODO

 - NumPy
 
    TODO

 - dask-ML
 
    TODO

### Dask issues and potential improvements (FIXME)

It was nice to see that we were not the only ones facing these kind of issues. In fact, almost all major issues mentioned at the workshop were shared among multiple attendees. Some of these were:
- Stability: in particular network stability, lack of replication of results, downscaling issues.
- Performance: lack of memory backpressure (OOME's are a thing), other issues involving suboptimal graph optimization  (e.g. task fusing).

> - Suboptimal work stealing.

Reviewer 2: Same as above: This feels a bit like counting problems, which is not bad per se. But I would like it if you could provide a bit more context on these problems. Maybe mention less, but spend a few sentences on those that you mention. Imho that would be more interesting.

Other issues mentioned involved indexing/partitioning, the lack of a multi-index and "non-functional" items such as documentation and examples.

The atmosphere of the workshop was great, people were eager to share their experiences with Dask and the ecosystem surrounding it. This provided for some interesting discussions.

### Working sessions (FIXME)

> Note: the following is purely a result of personal experience, I'm sure there are plenty of interesting things were done or talked about of which I am not aware.

Reviewer 2: I would probably leave this sentence out completely, as there is no information gain. If you want to introduce the topics you specifically discussed during the workshop, then just say it like that:
"During the workshop, I specifically worked on blahblahblah and did XYZ with Mrs. FamousDeveloper ..."

- Discussions on improving stability
- Distributed semaphore

> - Benchmarking work stealing for tasks with restrictions

Reviewer 2: Again: I would love to hear more details here, but maybe this is just me :)

Personally, I cannot say it enough times: I loved the working sessions. I learned a lot about Dask and distributed internals and made my first contributions to the project.

## Final remarks (FIXME)

> All in all, my colleagues and I found this workshop very valuable. We found this value in being able to share our issues and discuss potential ways of improvement. Lstening to other Dask users and developers gave us a wider perspective on the current status of Dask, the ecosystem around it, and what we could expect in the future.

Reviewer 2: Typo: "Lstenting"

Coincidentally, we also talked with other attendees of the workshop whom were interested in such a functionality.
This is something we did not forget; despite the jetlag, we left D.C. with a good chunk of the implementation necessary for a semaphore. This finally got [released](https://github.com/Dask/distributed/commit/2129b740c1e3f524e5ba40a0b6a77b239d4c1f94) with distributed 2.14.0.

### Interaction with the community
Being able to share our issues with the Dask community and discuss potential ways of improvement with expert users and core developers was extremely valuable. Additionally, this interaction gave us a wider perspective of the current status of Dask, the ecosystem around it and what we can expect from it in the future.

We were also reminded once again of the value of open-source software. 
There was a clear synergy in terms of intended functionality between [Kartothek](https://github.com/JDASoftwareGroup/kartothek) and other tools in the ecosystem such as Arrow/Parquet and Dask.dataframe I/O and partitioning.
Dask is driven by a wide community of skilled and passionate developers. 

> By committing to this community-driven project, we get to collaboratively shape the best software out there.

Reviewer 3: Noice!

---
layout: single
title:  "Karlsruhe to D.C. ― a Dask story"
date:   2020-03-24 20:00:00 +0100
tags: technology python data-engineering Dask distributed
header:
  overlay_image: assets/images/dask_icon.svg
  overlay_filter: 0.2
  show_overlay_excerpt: false
author: Lucas Rademaker & Nefta Kanilmaz
author_profile: true
---
# Karlsruhe to D.C. ― a Dask story

> Back in February 2020, we (Florian Jetter, Nefta Kanilmaz and Lucas Rademaker) travelled to the Washington D.C. metropolitan area to attend the first Dask developer conference. Our primary goal was to discuss Blue Yonder’s issues related to Dask with the attending developers and users. When we returned back to Germany, we had not only connected with many of the core developers in the Dask community, but also had an (almost finished) implementation of a distributed semaphore in our bags. 

Reviewer 3: Suggestions:
"... to attend the first ever Dask developer conference"
"... to Germany, ~~we had not only connected~~ not only had we connected..."
"... but ~~also had an (almost finished) implementation~~ had also (almost finished) an implementation of a distributed semaphore ~~in our bags.~~ [*unless* that's a `dask.bag` pun, in which case, please keep the original working]

```suggestion
Back in February 2020, we (Florian Jetter, Nefta Kanilmaz and Lucas Rademaker) travelled to the Washington D.C. metropolitan area to attend the first ever Dask developer conference. Our primary goal was to discuss Blue Yonder’s issues related to Dask with the attending developers and users. When we returned back to Germany, not only had we connected with many of the core developers in the Dask community, but we had also (almost finished) an implementation of a distributed semaphore. 
```

Author: > unless that's a dask.bag pun, in which case, please keep the original working

I thought it kinda was (@nefta-kanilmaz-jdas ?), I'll keep that for now

```suggestion
Back in February 2020, we (Florian Jetter, Nefta Kanilmaz and Lucas Rademaker) travelled to the Washington D.C. metropolitan area to attend the first ever Dask developer conference. Our primary goal was to discuss Blue Yonder’s issues related to Dask with the attending developers and users. When we returned back to Germany, not only had we connected with many of the core developers in the Dask community, but we had also (almost finished) an implementation of a distributed semaphore in our bags. 
```

> There is [a blog post](https://blog.dask.org/2020/04/28/dask-summit) giving an overview on and summary of the workshop talks. We want to share our experience around this workshop. 

Reviewer 3: "... a general overview ~~on~~ and summary of.... This blog post is a summary of _our_ experience of the workshop."

```suggestion
There is [a blog post](https://blog.dask.org/2020/04/28/dask-summit) giving a general overview and summary of the workshop talks. This blog post is a summary of _our_ experience of the workshop.. 
```

## Setting

> The three-day long workshop included sessions with short talks, as well as time slots for working on Dask-related issues and discussions.

Reviewer 3: "... as well as dedicated time slots...."

```suggestion
The three-day long workshop included sessions with short talks, as well as dedicated time slots for working on Dask-related issues and discussions.
```

> Talks covered a broad spectrum of topics including Dask-ML and usage of Dask in general data analysis, Dask deployment and infrastructure, and many more. All talks focused on answering these key questions: 

Reviewer 3: "The talks covered... and the usage of Dask... Dask deployment ~~and~~ & infrastructure ~~,~~ [no comma before and] and many more. All the talk seemed to focus on answering ~~these~~ three key questions:"

```suggestion
The talks covered a broad spectrum of topics including Dask-ML and the usage of Dask in general data analysis, Dask deployment & infrastructure and many more. All talks seemed to focus on answering three key questions: 
```

Author: why `&` over `and`?

- Who are Dask users and what are their use cases?
- What are current pain points and what needs to be fixed as soon as possible?

> - What is on the Dask user’s wish list?

Reviewer 3: "... ~~user's~~ users'..."

```suggestion
- What is on the Dask users' wish list?
```

> In other words, this was the chance for presenters to officially “rant” [*] about Dask, with experts having the experience and knowledge to help in the same room.

Reviewer 3: "~~In other words~~ In Matthew Rocklin's own words, this was ~~the chance~~ an opportunity...."

Indented formatting starts one character too soon: both square braces should be unitalicised.

```suggestion
"In Matthew Rocklin's own words, this was an opportunity for presenters to officially “rant” [*] about Dask, with experts having the experience and knowledge to help in the same room.
```

This exchange between users and developers enabled immediate fixes of minor problems, identifying synergies between different projects, as well as shaping the roadmap for future development of Dask. We feel that this is a successful model for driving the Dask open-source community.

> [*] quote Matthew Rocklin

Reviewer 3: The tenses in the quote appear to be messy. Is it a verbatim quote?

Author: @nefta-kanilmaz-jdas

Author: rendering issue: 
![image](https://user-images.githubusercontent.com/57136044/83024601-ecccb580-a02e-11ea-8745-1387e3f65bd3.png)

## The Blue Yonder way of using Dask

> We at Blue Yonder are still finishing our migration from a proprietary job scheduler to Dask.distributed. This is why when _we_ talk about _Dask_ in this post, we are mainly referring to _Dask.distributed_, as this is our main use case.

Reviewer 3: "... talk about ~~_Dask_~~ Dask in this post, we are ~~mainly referring~~ mostly referring to..."

```suggestion
We at Blue Yonder are still finishing our migration from a proprietary job scheduler to Dask.distributed. This is why when _we_ talk about Dask in this post, we are mostly referring to _Dask.distributed_, as this is our main use case.
```

Author: ```suggestion
We at Blue Yonder are still finishing our migration from a proprietary job scheduler to Dask.distributed. This is why when _we_ talk about _Dask_ in this post, we are mostly referring to _Dask.distributed_, as this is our main use case.
```

> We quickly realized at the workshop that Blue Yonder’s use case is almost unique in the Dask community.

Reviewer 3: "... in the entire Dask community."

```suggestion
We quickly realized at the workshop that Blue Yonder’s use case is almost unique in the entire Dask community.
```

> Florian Jetter opened the stage at the conference and presented the typical data flow for our machine learning products, the usage of Dask and Dask.distributed within this flow, and where we were currently facing issues.

Reviewer 3: "opened the stage" sounds colloquial. "opened proceedings" would read better.

 ```suggestion
Florian Jetter opened proceedings at the conference with a presentation on the typical data flow for our machine learning products, our usage of Dask and Dask.distributed within this flow and where we were currently facing issues.
```

> At Blue Yonder, we provide Machine Learning as a Service to our customers in retail. The data which our customers provide through an API is inserted into a relational database. The Machine Learning and prediction steps need a denormalized data format, which we provide in form of Parquet datasets in the Azure Blob Store. The resulting predictions are written back to Parquet files and offered to the customer through an API. 

Reviewer 3: ```suggestion
At Blue Yonder, we provide Machine Learning as a Service to our customers in retail. The data - which our customers provide through an API - is inserted into a relational database. The Machine Learning and prediction steps need a denormalized data format, which we provide in the form of Parquet datasets in the Azure Blob Store. The resulting predictions are written back to Parquet files and offered to customers through an API. 
```

The Data Engineering team leverages the distributed scheduler for parallel task execution during the data transformation steps between database and Parquet datasets. Most parts of these pipelines use Dask bags or delayed objects for map/reduce operations. Dask dataframes are especially useful when we reshuffle datasets with an existing partitioning to a differently partitioned dataset. We use [Kartothek](https://github.com/JDASoftwareGroup/kartothek) for this purpose.

> A lot of Dask users we met at the workshop also utilized Dask clusters for data-heavy calculations. However, the Blue Yonder use case remained special: most other users were working in an R&D-like environment. Their computations are not running in comparable production environments. Most importantly, these users had no service level agreements with customers about, for example, delivery times for predictions. It seems that Blue Yonder puts higher requirements on the stability and robustness of the distributed scheduler than any other of the users present at the conference.

Reviewer 3: ```suggestion
A lot of Dask users we met at the workshop also utilized Dask clusters for data-heavy calculations. However, the Blue Yonder use case remains special: most other users are working in an R&D-like environment and their computations are not running in comparable production environments. Most importantly, these users have no service level agreements with customers: SLAs, for example, for delivery times for predictions. From our interactions with users, it became apparent that Blue Yonder puts far higher requirements on the stability and robustness of the distributed scheduler than any other users present at the conference.
```

> ## Issues we have encountered with Dask and potential improvements

Reviewer 3: The title doesn't summarize the content, IMHO.

```suggestion
## Issues with Dask that others have encountered too
```

Author: But with that suggestion we remove the notion that _we_ have encountered the issues also, is there a compromise?

Reviewer 3: The "too" at the end implies that you've faced it. If you want to call it out more explicitly then maybe something like:
```suggestion
## We are not alone - Dask issues encountered by others too 
```
or
```suggestion
## Dask issues faced by us and others too

Author: @nefta-kanilmaz-jdas : _no_

> In these discussions we talked about a lot of topics, some of which have been or still are big problems for us. We encountered community members who had similar experiences. Below are some of the "rants" we want to share.

Reviewer 3: In _which_ discussions?

```suggestion
During discussions with conference attendees we touched upon some topics that have been (or still are) big problems for us and encountered other community members with similar experiences. Below are some of the "rants" we'd like to share.
```

### Distributed stability

> We migrated some of our largest data pipelines to Dask in the last months of 2019. For these pipelines, we started to observe significant instability during the computation of the Dask graph. We hadn't seen this issue in any of our previous pipelines running on distributed. Our team contributed several patches upstream in order to resolve these issues on our side. 

Reviewer 3: ```suggestion
We migrated some of our largest data pipelines to Dask in the last months of 2019. For these pipelines, we started to observe significant instability during the computation of the Dask graph. We hadn't seen this issue in any of our previous pipelines running on distributed. To resolve these issues on our side, our team ended up contributing several patches upstream. 
```

Author: "physicists English" according to @nefta-kanilmaz-jdas, I don't have strong feelings about it tho

Reviewer 3: :P Not every reader is necessarily going to be a physicist, though, thus this suggestion.

> During the working sessions, a number of conversations took place related to the stability of the distributed scheduler, involving Dask developers and users.

Reviewer 3: ```suggestion
During the working sessions, a number of conversations related to the stability of the distributed scheduler took place.
```

> One idea which came up to increase the overall robustness of distributed was _replication of task results_. The idea is the following: if a worker which is holding a task result is shut down, but the task result is replicated, then we do not need to re-compute parts of the graph again.

Reviewer 3: ```suggestion
One idea which came up, to increase the overall robustness of distributed, was _replication of task results_. The idea is as follows: if a worker which is holding a task result is shut down, replicating the task result will allow certain parts of the graph to not need to be re-computed.
```

As of the time of this writing, we have not had a chance to actively work on something like this yet, but it is something that we keep in mind.

### Performance and graph optimization
While monitoring the execution of one of our Dask production pipelines, we observed that the amount of memory consumed by the workers was significantly higher than one would expect in an ideal scenario where tasks are executed in a way which minimizes memory usage of the workers.

> When investigating the Dask.optimization module, we saw that code to optimize memory usage was already there. In practice, however, the graphs are not executed in such an order because of other constraints during execution.

Reviewer 3: "In practice, however, the graphs are not executed in such an order because of other constraints during execution" - what does such an order here refer to?

Author: Optimal order in terms of memory usage

Author: ```suggestion
When investigating the Dask.optimization module, we saw that code to optimize memory usage was already there. In practice, however, the graphs are not executed in the optimal order with regards to memory usage because of other constraints during execution.
```

A Dask user at the workshop facing this issue told us that they worked around this by injecting dummy dependencies into Dask graphs. These dependencies acted as "choke-holds" for certain types of tasks, in order to improve the memory usage during execution.

> This lack of optimization, also referred to as memory back-pressure, does not only impact memory usage but also increases resource consumption. We would greatly benefit from an implementation which addresses this issue.

Reviewer 3: ```suggestion
This lack of optimization, also referred to as memory back-pressure, not only impacts memory usage but also increases resource consumption. We would greatly benefit from an implementation which addresses this issue.
```

## Looking back

### Results from working sessions
Lucas had the pleasure to collaborate with John Lee on writing down a benchmark to enable [work stealing for tasks with restrictions](https://github.com/Dask/distributed/pull/3069). He also got valuable input from Tom Augspurger on dask internals while debugging a Dask.dataframe bug involving categoricals.

However, arguably the most useful work we did was the implementation of a semaphore in Dask.distributed.

> We need to be able to rate-limit the access of the computation clusters to certain resources such as production databases. Migrating workflows which were still heavily dependent on accessing the database to the distributed scheduler was not possible without a semaphore implementation.

Reviewer 3: ```suggestion
Migrating workflows which were still heavily dependent on accessing the database to the distributed scheduler was not possible without a semaphore implementation. For this, we needed to be able to rate-limit the access of the computation clusters to certain resources such as production databases. 
```

Author: I think the current order of sentences is still valuable as it gives the reason/cause first and then the solution.

Reviewer 3: Alright, if that's the original intention, then I'd suggest:
```suggestion
Migrating workflows, which were still heavily dependent on accessing resources like production databases, to the distributed scheduler was not possible without being able to rate-limit the access of computation clusters to them. For this, we needed a semaphore implementation.

We need to be able to rate-limit the access of the computation clusters to certain resources such as production databases. Migrating workflows which were still heavily dependent on accessing the database to the distributed scheduler was not possible without a semaphore implementation.
```

Author: 👍

> By working on the semaphore during the workshop, we unblocked a Blue Yonder product team in its migration to Dask.

Reviewer 3: This entire sentence is premature + unnecessary, since you say it again two sentences down: "This is something we did not forget;..."

I'd suggest modifying the following three sentences. Something like:

```suggestion
We talked to other workshop attendees who were also interested in a semaphore functionality and during the course of the workshop, actually started implementing it. We finally left DC with a good chunk of the implementation already in place and despite the jetlag, continued working on it once we got back to Karlsruhe. This finally got released...."
```

Author: you are omitting `we unblocked a Blue Yonder product team in its migration to Dask` here though

Reviewer 3: Fair point. Updated suggestion:
```suggestion
We talked to other workshop attendees who were also interested in a semaphore functionality and during the course of the workshop, actually started implementing it. We finally left DC with a good chunk of the implementation already in place and despite the jetlag, continued working on it once we got back to Karlsruhe. This finally got [released](https://github.com/Dask/distributed/commit/2129b740c1e3f524e5ba40a0b6a77b239d4c1f94) with distributed 2.14.0 and enabled us to unblock a Blue Yonder product team in its migration to Dask.
```

Author: We'll keep it for now

Coincidentally, we also talked with other attendees of the workshop whom were interested in such a functionality.
This is something we did not forget; despite the jetlag, we left D.C. with a good chunk of the implementation necessary for a semaphore. This finally got [released](https://github.com/Dask/distributed/commit/2129b740c1e3f524e5ba40a0b6a77b239d4c1f94) with distributed 2.14.0.

### Interaction with the community

> Being able to share our issues with the Dask community and discuss potential ways of improvement with expert users and core developers was extremely valuable. Additionally, this interaction gave us a wider perspective on the current status of Dask, the ecosystem around it, and what we can expect in the future.

Reviewer 3: ```suggestion
Being able to share our issues with the Dask community and discuss potential ways of improvement with expert users and core developers was extremely valuable. Additionally, this interaction gave us a wider perspective of the current status of Dask, the ecosystem around it and what we can expect from it in the future.
```

> We were also reminded once again by the value of open-source software. 

Reviewer 3: ```suggestion
We were also reminded once again of the value of open-source software. 
```

There was a clear synergy in terms of intended functionality between [Kartothek](https://github.com/JDASoftwareGroup/kartothek) and other tools in the ecosystem such as Arrow/Parquet and Dask.dataframe I/O and partitioning.

> We are sharing responsibility of Dask with the community; by committing to this project we are accessing a broad pool of developers.

Reviewer 3: This sentence and the one below - I do not understand the intention. What exactly do you want to express?

Author: Mostly pro-OSS propaganda/appreciation

Author: Would you phrase this differently?

Reviewer 3: > Mostly pro-OSS propaganda/appreciation
Lol
> Would you phrase this differently?
Erm.. I'd suggest something like:

```suggestion
Dask is owned and driven [hope this is true] by a wide community of skilled and passionate developers. For us, therefore, committing to Dask implies not only that we get access to a great piece of software but also that we get to collaboratively helps shape the future of the project. 
```

Note, that that's a first draft suggestion. Will post further suggestions here if I come up with better ones.

Reviewer 4: It's owned and driven primarily by Matt Rocklin AFAIK, so that is not entirely accurate

Reviewer 3: Lol, okay. Will think of how to re-rephrase.

Reviewer: I am sure, Matthew likes to read, that there is a strong community instead of a one man show...no matter what the truth is...we should do him the favor and say it is a strong community, so I like Kishs suggestion

Reviewer 4: ```suggestion
Dask is driven by a wide community of skilled and passionate developers. For us, therefore, committing to Dask implies not only that we get access to a great piece of software but also that we get to collaboratively help shape future developments. 
```

In this sense, to be able to get the best software out there, collaborating with the community is a necessity. 

### Final remarks

> We were grateful to have the opportunity to attend this workshop.

Reviewer 3: ```suggestion
We are grateful to have had the opportunity to attend this workshop.
```

---
layout: single
title:  "Tales from D.C.: Dask Developer Workshop 2020"
date:   2020-03-24 20:00:00 +0100
tags: technology python data-engineering Dask distributed
header:

>   overlay_image: assets/images/tech_gear_banner.jpg

Reviewer 4: Sure, will try something

This exchange between users and developers enabled immediate fixes of minor problems, identifying synergies between different projects, as well as shaping the roadmap for future development of Dask. We feel that this is a successful model for driving the Dask open-source community.

[*] quote Matthew Rocklin

## The Blue Yonder way of using Dask

We quickly realized that Blue Yonder’s use case is almost unique in the Dask community.

Florian Jetter opened the stage at the conference and presented the typical data flow for our machine learning products, the usage of Dask and Dask.distributed within this flow, and where we were currently facing issues.
 
At Blue Yonder, we provide Machine Learning as a Service to our customers in retail. The data which our customers provide through an API is inserted into a relational database. The Machine Learning and prediction steps need a denormalized data format, which we provide in form of Parquet datasets in the Azure Blob Store. The resulting predictions are written back to Parquet files and offered to the customer through an API. 

> The Data Engineering team leverages the distributed scheduler for parallel task execution during the data transformation steps between database and Parquet datasets. Most parts of these pipelines use Dask bags or delayed objects for map/reduce operations. Dask dataframes are especially useful when we reshuffle datasets with an existing partitioning to a differently partitioned dataset.

Reviewer 4: - [ ] link ktk

> A lot of Dask users we met at the workshop also utilized Dask clusters for data-heavy calculations. However, the Blue Yonder use case remained special: most other users were working in an R&D-like environment. Their computations are not running in comparable production environments. Most importantly, these users had no service level agreements with customers about, for example, delivery times for predictions. It seems that Blue Yonder puts higher requirements on the stability and robustness of the distributed scheduler than any other of the users present at the conference.

Reviewer 2: >In particular, a few people mentioned this also being a problem but its not that critical as it is to us. Also some suggested that they think of moving stuff more and more to production.

Do you read this section as "All others are R&D and we are the only ones doing real production stuff"? 

What I read from your proposed changes:
* You would put more emphasis on the SLAs, also explaining them in even simpler terms
* You would stress out that we were the first ones using distributed in this way and, with that, were "pavong the road" for other users

Is that about right?

> ## Issues we have encountered with Dask and potential improvements

Reviewer 4: Do we really mention what the solutions _shall_ be? AFAIK we talk about fixes for past issues (but don't go into detail) and _potential_ fixes for current issues

### Distributed stability
We started migrating one of our largest data pipelines to Dask in the last months of 2019. With these newly migrated pipelines, we started to observe significant instability during the computation of the Dask graph.
One of the issues that we identified was a scenario wherein the scheduler saw a worker as being in an "invalid" state and, as a result, killed it, although

> the worker had been running fine up until that point. This issue was related to a temporary loss of network connectivity between worker and scheduler. The team at Blue Yonder contributed several patches to increase the robustness of the distributed scheduler with regard to network stability.

Reviewer 4: Just to be clear, you're suggestion is to include an explanation about spurious broken connections in this paragraph?

We also observed critical behaviour of the distributed scheduler when the cluster is being downsized. Downscaling/downsizing refers to the process of workers being removed from an active cluster.
Downsizing of a cluster during computation can result in a re-computation of large parts of the graph. This occurs when a worker that is holding a task result is being shut down. This task will then need to be re-computed including all necessary dependencies, leading to significantly increased computation time or possible never-ending executions.

> The Blue Yonder Core Platform team implemented several work-arounds to allow for graceful downscaling, but this has not yet been addressed in a generic way in the distributed scheduler.

Reviewer 4: Does it still make sense to talk about this then?

>  

Reviewer 4: ```suggestion
```

One idea which may help mitigate the above issues is _replication of some task results_. The idea is the following: if a worker which is holding a task result is shut down, but the task result is replicated, then we do not need to re-compute parts of the graph again.
One could think about replicating task results for tasks that qualify some criteria such as long running time or small task size.

### Performance and graph optimization
While monitoring the execution of one of our Dask production pipelines.
The amount of memory consumed by the workers was significantly higher than one would expect in an ideal scenario where tasks are executed in a way which minimizes memory usage of the workers.
When investigating the Dask.optimization module, we saw that code to optimize memory usage was already there. In practice, however, the graphs are not executed in such an order because of other constraints during execution.
A Dask user at the workshop facing this issue told us that they worked around this by injecting dummy dependencies into Dask graphs. These dependencies acted as "choke-holds" for certain types of tasks, in order to improve the memory usage during execution.

> This lack of optimization referred to as memory back-pressure is currently not a priority for Blue Yonder, as increasing worker memory served as a work-around. There still may be value in exploring this further in the future.

Reviewer 4: - [ ] re-phrase

## Looking back

### Results from working sessions
Lucas had the pleasure to work with [John Lee](https://github.com/jlee3) on writing down a benchmark to enable [work stealing for tasks with restrictions](https://github.com/Dask/distributed/pull/3069), as well as with [Tom Augspurger](https://github.com/TomAugspurger) on a Dask.dataframe bug involving categoricals (PR still pending).

However, arguably the most useful work we did was the implementation of a semaphore in Dask.distributed.
We need to be able to rate-limit the access of the computation clusters to certain resources such as production databases. Migrating workflows which were still heavily dependent on accessing the database to the distributed scheduler was not possible without a semaphore implementation.
By working on the semaphore during the workshop, we unblocked a Blue Yonder product team in its migration to Dask.

>  Coincidentally, we also talked with other attendees of the workshop whom were interested in such a functionality.

Reviewer 2: > I wouldn't call the semaphore a result of the workshop since we were doing it anyhow. 
Technically right, but does the reader of this blog benefit from knowing this difference?

I would say, the workshop accelerated work on the semaphore significanlty and it is something that blocked LSF. We can add the angle about community requests, it fits in nice here.

>  Coincidentally, we also talked with other attendees of the workshop whom were interested in such a functionality.

Author: fixed typo

---
layout: single
title:  "Tales from D.C.: Dask Developer Workshop 2020"

> date:   2020-03-24 20:00:00 +0100

Reviewer 4: change date before publishing

tags: technology python data-engineering Dask distributed
header:
  overlay_image: assets/images/tech_gear_banner.jpg
  overlay_filter: 0.2
  show_overlay_excerpt: false

> author: Lucas Rademaker, Nefta Kanilmaz

Reviewer 4: Multiple authors: will this work?

Author: It's not supported unfortunately
